azure/LongRunningOperations.md

```diff
@@ -0,0 +1,69 @@
+# Long-Running Operations
+
+Long-running operations are an API design pattern that should be used when the processing of
+an operation may take a significant amount of time -- longer than a client will want to block
+waiting for the result.
+Azure allows for two forms of this design pattern: resource-based long-running operations (RELO),
+which is the preferred pattern, and stepwise long-running operations.
+
+In both patterns, the processing of the operation is initiated by one API call and the client
+obtains the results of the operation from a subsequent API call.
+Here we illustrate the sequence of API calls involved in each of these patterns.
+
+## Resource-based long-running operations
+
+In the RELO pattern, the resource that is the target of the operation contains a `status` field
+that holds the status of an outstanding or last completed operation.
+This means that the client can use a standard "get" operation on the resource to determine the
+status of an operation it initiated.  The flow looks like this:
+
+![The RELO flow](./relo.jpg)
+
+1. The client sends the initial request to the resource to initiate the long-running operation.
+This initial request could be a PUT, PATCH, POST, or DELETE method.
+
+2. The resource validates the request and initiates the operation processing.
+It sends a response to client with a `200-OK` HTTP status code (or `201-Created` if the operation
+is a create operation) and a representation of the resource where the `status` field is set
+to a value indicating that the operation processing has been started.
+
+3. The client then issues a GET request to the resource to determine if the operation processing
+has completed.
+
+4. The resource responds with a representation of the resource. While the operation is still being
+processed, the status field will contain a "non-terminal" value, like "processing".
+
+5. After the operation processing has completed, a GET request from the client will receive a response
+where the status field contains a "terminal" value -- `succeeded`, `failed`, or `cancelled` --
+that indicates the result of the operation.
+
+## Stepwise long-running operations
+
+In the stepwise LRO pattern, the status and results of the operation are encapsulated into an
+operation-results object that is distinct from the target resource and specific to the individual
+operation request.  Here's what the stepwise LRO pattern looks like:
+
+![The stepwise LRO flow](./stepwise.jpg)
+
+1. The client sends the initial request to initiate the long-running operation.
+As in the RELO pattern, the initial request could be a PUT, PATCH, POST, or DELETE method.
+
+2. The resource validates the request and initiates the operation processing.
+It sends a response to client with a `202-Accepted` HTTP status code.
+Included in this response is an `Operation-location` response header with the absolute URL of an
+operation-results endpoint for this specific operation.
```

Comment:
To me, this question comes up: Why does the client need the operation-location URL if the operation-result URL is well-known? I think we're over complicating.
Now, you might say because a POST to initiate the operation has to return the operation ID so it can be used to look up the operation in the operation-results collection but then why don't we say that operation should just be initated via a PUT to the operation-results collection and then the ID is passed by the client and so there is no need for the operation-location header at all. And, this makes starting the operation idempotent too.

Reply:
I think this confusion comes from my choice of naming.  I was calling the URL returned in the `Operation-location` header the "operation-results" URL.  Perhaps calling it the "operation-location" URL is better -- or maybe just "operation URL"?